Ability to Change View Scroll Speed

adakitesystems opened the pull request:
This pull request implements feature #1355: Ability to Change View Scroll Speed

**Details:**
- Possible internal scroll speed values are ranged 1 to 435 inclusive (default=12) which are the results from an exponential growth equation where `x` is an integer between 1 and 100 inclusive. `x` is set via a scroll bar in the "Options" window. The equation was selected based on local gameplay testing, personal preferences, and scrolling behavior personally experienced playing other games.
- The equations used are `y = |(6/5)^(x/3)|` and `x = 3log(y) / log(6/5)`. Suggestions on equation alternatives are welcome.

**Comments:**
I am personally fairly satisfied with the result. However, because of the exponential nature of the equation, there is a visual awkwardness with very low values (see screenshots) which I dislike. In the attached screenshots, both scroll bar positions produce the internal value of 1. If the window is closed as seen in screenshot (1), it will be opened appearing as screenshot (2). The internal value is still correctly loaded/saved as 1, but it is not visually represented accurately as it is with higher values.

Thoughts/comments/suggestions/criticism welcome.

**Screenshots**:
1)
![inaccurate1](https://user-images.githubusercontent.com/22711643/34801275-69147b22-f62d-11e7-96ae-67508745d9a8.png)
2)
![inaccurate2](https://user-images.githubusercontent.com/22711643/34801276-6974eb7e-f62d-11e7-98a0-bd1d00fe156c.png)

adakitesystems commented:
There seems to be a widget bug with the `Invert right mouse dragging` widget when switching to any tab BEYOND the "Controls and interface" (the first 4 tabs do not trigger the bug) and is **only present in my commit**. Possibly some lack of invalidation when I added the label and scroll bar widgets? Any ideas? I'm looking into this.
![control_options_bug](https://user-images.githubusercontent.com/22711643/34803564-92c6d8a0-f639-11e7-927d-11dab0ae66b8.png)

AaronVanGeffen commented:
> There seems to be a widget bug with the `Invert right mouse dragging` widget when switching to any tab BEYOND the "Controls and interface" tab (the first 4 tabs do not trigger the bug) and is **only present in my commit**. Possibly some lack of invalidation when I added the label and scroll bar widgets? Any ideas? I'm looking into this.

Your label and scroll widgets overlap horizontally. That's probably why.

Personally, I'm not too keen on using scrollbar widgets as sliders. I know it's done in the music tab, too, but it's not ideal — especially not between these checkboxes. It might look better if you move both the edge-scrolling checkbox and your new widgets down a little, just below the 'Zoom to cursor' checkbox?

Another option would be to do away with the idea of a slider entirely, and just replace the edge scrolling checkbox with a dropdown that lets you choose between several tiers.

adakitesystems commented:
Thanks for the quick response. I tried some things but had no luck.

> Your label and scroll widgets overlap horizontally. That's probably why.

I noticed that the checkboxes and scroll bars from the audio tab overlap horizontally as well but they seem to have no problem. Nevertheless, I did shorten the label to end at pixel 145 instead of 299 and that didn't seem to change anything. However, when I completely removed the label code and instead of `Invert right mouse dragging` being changed into a bugged half-dropdown-looking box, `Zoom to cursor position` was now experiencing the same problem. 

I also noticed that only the "Misc" tab seems to create problems for the "Controls and interface" tab with the new widgets. If you click any other tab before clicking "Controls and interface", there is no problem. I'll have to continue analyzing how to properly add widgets.

> It might look better if you move both the edge-scrolling checkbox and your new widgets down a little, just below the 'Zoom to cursor' checkbox?

I was aiming to keep the scroll speed bar close to the existing `Scroll view when pointer at screen edge` checkbox since those two are closely related. Maybe both of those widgets should be moved to the bottom of the group box before the `Shortcut keys` button?

> Another option would be to do away with the idea of a slider entirely, and just replace the edge scrolling checkbox with a dropdown that lets you choose between several tiers.

I disagree with this approach because of the nature of the option and its possible values. This new feature is similar to sound, mouse pointer sensitivity, window sizes, etc. I don't think tiers would be a good fit just as tiers for sound (Soft, Normal, Loud) or mouse pointer sensitivity (Slow, Fast, Fastest) would not be a good fit.

Gymnasiast commented:
Sorry for coming off harsh, but I'm not sure what makes scroll speed warrant a feature. (Yes, I know some people like an option for _everything_.) There is always right mouse button scrolling if it's not quick/slow enough.

adakitesystems commented:
It's no problem at all and I understand the rationale for rejecting this PR. I am also not deterred by criticism (whether harsh or not) as long as its constructive and/or backed up by an argument like you have presented.

I know of a few popular(ish) RTS games which have scroll speed in their options and is used by the majority of their playerbases for navigating around the view space a bit more quickly. The options window has a slider as well for adjusting the value. Those games even have minimaps which allow jumping to any section of the map instantly. Players use the minimap and also customize their scroll speed. Scroll speed is not a novel idea or a customization specific to RCT.

I understand wanting to keep the amount of customizations low to avoid clutter. Don't worry about me. If you would like the feature, it's' pretty much ready to go (except for the widget). If not, then no hard feelings :)

adakitesystems commented:
If you're interested in which other games this option is present, [here are some screenshots](https://imgur.com/a/sFjV9) of scroll speed as a slider in StarCraft II, WarCraft II, and StarCraft: Brood War. I believe there is a high chance scroll speed is also included in games such as DotA 2, League of Legends, and a lot other games in which a player moves the mouse to edge of the screen in order to shift the screen position.

marijnvdwerf commented:
I wouldn't want an in-game option to change to scroll speed (as it would only further complicate the options dialog), but I'd argue that the application of scroll speed becomes a lot more obvious because of this PR. `edge_scrolling_speed` has a lot more meaning than `12`.

Gymnasiast commented:
Ok, we discussed this and this is the result:

Adding the config option is fine, but don't add it to the Options window. This effectively means just dropping the second commit, rebasing and addressing CI errors, if any.

adakitesystems commented:
Great! I have dropped the first commit, rebased, and there appear to be no CI errors. Please let me know if this PR requires further discussion/modification. Thanks, everyone!

rwjuk commented:
@adakitesystems Merged - thanks for your contribution! :)
